[debugger plugin] Open source debugger plugin: with health pills and numeric alerts

caisq opened the pull request.

caisq commented:
@chihuahua, @jart, @wchargin

wchargin commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+        tf.logging.info("Created directory for debugger data: %s",
+                        debugger_directory)
+      except tf.OpError as e:
+        tf.logging.error(
```
Is this not fatal?

caisq replied:
Right. Done.

wchargin commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+
+        try:
+          # Use the data to initialize the registry.
+          initial_data = json.loads(backup_data_json_string)
```
You should be able to replace this and the definition above with `initial_data = json.load(backup_file)`.

caisq replied:
Done.

wchargin commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+    Args:
+      alert: The alert to be registered.
+    """
+    self._numerics_alert_lock.acquire()
```
why not `with self._numerics_alert_lock: self._numerics_alert_registry.register(alert)`?

caisq replied:
Good point. Done.

wchargin commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+      A list of `numerics_alert.NumericsAlertReportRow`, sorted in ascending
+        order of first_timestamp.
+    """
+    self._numerics_alert_lock.acquire()
```
why not `with self._numerics_alert_lock: return self._numerics_alert_lock_registry.report()`?

caisq replied:
Done.

wchargin commented on tensorboard/plugins/debugger/debugger_server_test.py:
```diff
@@ -0,0 +1,260 @@
+    self.events_written = events_output_list
+
+  def dispose(self):
+    """Does nothing. This implementation creates no file."""
```
I know this is technically valid, but can you add a `pass` statement here, or will pylint complain?

caisq replied:
pylint is happy with this.

wchargin commented on tensorboard/plugins/debugger/debugger_server_test.py:
```diff
@@ -0,0 +1,260 @@
+    value = event.summary.value.add()
+    value.tag = node_name
+    value.node_name = "%s:%d:%s" % (node_name, output_slot, debug_op)
+    value.tensor.dtype = tf.float64.as_datatype_enum
```
`tf.make_tensor_proto`!

caisq replied:
Done.

wchargin commented on tensorboard/plugins/debugger/debugger_server_test.py:
```diff
@@ -0,0 +1,260 @@
+    for expected_event, gotten_event in zip(expected, gotten):
+      self.assertEqual(expected_event.summary.value[0].node_name,
+                       gotten_event.summary.value[0].node_name)
+      self.assertEqual(
```
`tf.make_tensor_proto`, especially in this function, and especially especially on the assertion following this one!

caisq replied:
Done.

wchargin replied:
GitHub doesn't think so—are you sure that you pushed this change?

wchargin replied:
Not sure what the status of this is. The code that I want to see is of the form:
```py
self.assertAllClose(
    tf.make_ndarray(expected_event.summary.value[0].tensor),
    tf.make_ndarray(actual_event.summary.value[0].tensor))
```
It is **not safe** to assume that if two tensor protos are different then they represent different tensors. This is true both in theory and in practice: for instance, if [this function in `scalar/summary_test.py`](https://github.com/tensorflow/tensorboard/blob/615eccdaa5119d2c3aaf4870f5882fa2e606b47c/tensorboard/plugins/scalar/summary_test.py#L39) is replaced with the identity function, then the test **fails** on TensorFlow 1.3.

wchargin commented on tensorboard/plugins/debugger/events_writer_manager.py:
```diff
@@ -0,0 +1,239 @@
+
+# A regex used to match the names of debugger-related events files.
+_DEBUGGER_EVENTS_FILE_NAME_REGEX = re.compile(
+    "^" + re.escape(DEBUGGER_EVENTS_FILE_STARTING_TEXT) + "\\.\\d+\\.(\\d+)")
```
Please replace the last string with the raw string literal `r'\.\d+\.(\d+\)`, and also the first one with `r'^'` for consistency.

caisq replied:
Done.

wchargin commented on tensorboard/plugins/debugger/events_writer_manager.py:
```diff
@@ -0,0 +1,239 @@
+
+# By default, we check whether the current events file has exceeded the file
+# size cap every this many steps.
+_DEFAULT_CHECK_THIS_OFTEN = 1e5
```
This name could be improved: it should mention _what_ is being checked (the events file size), and preferably that the unit is "steps," too (but that's fine to omit if it's too verbose). Likewise with arguments/fields of the ~same name below.

caisq replied:
Done.

wchargin commented on tensorboard/plugins/debugger/events_writer_manager_test.py:
```diff
@@ -0,0 +1,435 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests the Tensorboard debugger data plugin."""
```
caps: Tensor**B**oard

caisq replied:
Done.

wchargin commented on tensorboard/plugins/debugger/events_writer_manager_test.py:
```diff
@@ -0,0 +1,435 @@
+from werkzeug import test as werkzeug_test
+from werkzeug import wrappers
+
+# pylint: disable=ungrouped-imports, wrong-import-order
```
Again, why?

caisq replied:
See my answer above.

wchargin commented on tensorboard/plugins/debugger/events_writer_manager_test.py:
```diff
@@ -0,0 +1,435 @@
+
+class DebuggerPluginTest(tf.test.TestCase):
+
+  def setUp(self):
```
This large `setUp` method, and much of the rest of this class, is duplicated across files. So: (1) all of my comments on the other file apply here, and (2) what do you think about extracting it to a base `DebuggerTestCase(tf.test.TestCase)` class? At a glance, it looks like hundreds of lines, so it sounds worth it to me.

caisq replied:
Sounds good. Done.

wchargin commented on tensorboard/plugins/debugger/numerics_alert.py:
```diff
@@ -0,0 +1,338 @@
+  value = event.summary.value[0]
+  debugger_plugin_metadata_content = None
+  if value.HasField("metadata"):
+    for plugin_data in value.metadata.plugin_data:
```
Again, make sure that https://github.com/tensorflow/tensorflow/pull/11952 is on your radar.

caisq replied:
Yep. @chihuahua let me know. I'll fix the tests once those changes reach tensorflow nightly.

wchargin replied:
Great.

wchargin commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+    # We remove the constants.DEBUG_NUMERIC_SUMMARY_SUFFIX from the end of the
+    # watch name because it is not distinguishing: every health pill entry ends
+    # with it.
+    node_name_and_output_slot = watch_key.rstrip(
```
(Looks like the implementation that I proposed is used elsewhere (`numerics_alert.py`); further evidence in favor of switching.)

caisq replied:
Good catch. Done.

wchargin commented on tensorboard/plugins/debugger/session_debug_test.py:
```diff
@@ -0,0 +1,339 @@
+                               debug_urls=[self._debug_url])
+
+          sess.run(a.initializer, options=run_options)
+          poll_success = True
```
Might as well just `return True` here and `return False` outside the `for` loop. Less state.

caisq replied:
Done.

wchargin commented on tensorboard/plugins/debugger/session_debug_test.py:
```diff
@@ -0,0 +1,339 @@
+              health_pills[debug_key].append(
+                  tf_debug.load_tensor_from_event(event_read))
+      except StopIteration:
+        break
```
You can simplify this by instead using `next_event = next(reader, None); if next_event is None: break; event_read.ParseFromString(next_event)` above. This limits the scope of the `try` block—e.g., you will not silently hide errors in `tf_debug.load_tensor_from_event`.

(I'm assuming from context that `None` is not a valid event. Otherwise: `sentinel = object(); next_event = next(reader, sentinel); if next_even is sentinel: break;`.)

caisq replied:
Done. Thanks for the suggestion!

wchargin commented on tensorboard/plugins/debugger/session_debug_test.py:
```diff
@@ -0,0 +1,339 @@
+    for debug_key in debug_key_to_tensors:
+      tensors = debug_key_to_tensors[debug_key]
+      for i, tensor in enumerate(tensors):
+        self.assertAllClose(
```
Huh, TIL about `tf.test.TestCase.assertAllClose` as an alternative to `np.testing.assert_allclose`. Nice.

caisq replied:
Ack.

wchargin commented:
Logistics:
  * What is `[tdp]`? (Tests don't pass, maybe? Why not?)
  * Please add a commit description that (suggestions follow) briefly describes this plugin, and specifically describes any non-obvious implications. (Like…why don't tests pass?) I also appreciate when PRs include test plans, but not everyone does this.

caisq commented on tensorboard/BUILD:
```diff
@@ -62,6 +63,14 @@ ts_web_library(
 )
 
 py_library(
+    name = "expect_mock_installed",
```
It just occurred to me that mock is available as tf.test.mock. Replaced this with that. Done.

wchargin replied:
Great; good solution.

…so we can remove this now, yes?

caisq replied:
Yes. Good catch. Done.

caisq commented on tensorboard/BUILD:
```diff
@@ -70,6 +79,14 @@ py_library(
 py_library(
+    name = "expect_portpicker_installed",
+    # This is a dummy rule used as a portpicker dependency in open-source.
+    # We expect numpy to already be installed on the system, e.g. via
```
Done.

caisq commented on .travis.yml:
```diff
@@ -42,6 +42,9 @@ before_install:
     esac
 
     pip install pylint
+    pip install futures==3.1.1
```
Done. It is actually a dependency of grpc_debug_server, which comes from the tensorflow package and is used by the debugger plugin. See:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/debug/lib/grpc_debug_server.py#L26

wchargin replied:
(You say "done" but it's still here—perhaps you forgot to commit this change?)

caisq replied:
Oops. I thought I removed it but didn't. Done (pending git push).

caisq commented on tensorboard/plugins/debugger/debugger_plugin.py:
```diff
@@ -0,0 +1,520 @@
+from __future__ import print_function
+
+import collections
+from collections import namedtuple
```
Done.

caisq commented on .travis.yml:
```diff
@@ -42,6 +42,9 @@ before_install:
     pip install pylint
+    pip install futures==3.1.1
+    pip install grpcio==1.4.0
+    pip install portpicker==1.1.1
```
OK. Done. Using port 4243 here.

wchargin replied:
The dependency is okay with me if(f) it's okay with @jart.

caisq commented on tensorboard/plugins/debugger/debugger_plugin.py:
```diff
@@ -0,0 +1,520 @@
+    self._debugger_data_server_grpc_port = debugger_data_server_grpc_port
+
+  def get_plugin_apps(self):
+    """Obtains a mapping between routes and handlers. Stores the logdir.
```
Updated the doc string.

caisq commented on tensorboard/plugins/debugger/debugger_plugin.py:
```diff
@@ -0,0 +1,520 @@
+      A mapping between routes and handlers (functions that respond to
+      requests).
+    """
+    # Start a debugger data server if a port is specified, and one has not been
```
Done.

caisq commented on tensorboard/plugins/debugger/debugger_plugin_test.py:
```diff
@@ -0,0 +1,436 @@
+from werkzeug import test as werkzeug_test
+from werkzeug import wrappers
+
+# pylint: disable=ungrouped-imports, wrong-import-order
```
There are some differences in the external and internal pylint checks. The internal one expects `import tensorflow as tf` and `from tensorflow.core.debug import debugger_event_metadata_pb2` to be in two separate groups, whereas the external one expect them to be together.

wchargin replied:
I see. Unfortunate, but I agree with your solution.

caisq commented:
@wchargin: [tdp] stands for [tensorboard debugger plugin]. :) Now I see it can be a little confusing. I'll think about alternative acronyms.

caisq commented on tensorboard/plugins/debugger/debugger_plugin.py:
```diff
@@ -0,0 +1,520 @@
+      A werkzeug BaseResponse object.
+    """
+    if request.method != 'POST':
+      tf.logging.error(
```
Done. Same below.

caisq commented on tensorboard/plugins/debugger/debugger_plugin.py:
```diff
@@ -0,0 +1,520 @@
+    try:
+      node_names = json.loads(jsonified_node_names)
+    except Exception as e:  # pylint: disable=broad-except
+      # Different JSON libs raise different exceptions, so we just do a
```
Agreed. TODO item added.

caisq replied:
Agreed. TODO item added.

caisq commented on tensorboard/plugins/debugger/debugger_plugin_test.py:
```diff
@@ -0,0 +1,436 @@
+
+  def setUp(self):
+    # Populate the log directory with debugger event for run '.'.
+    self.log_dir = tempfile.mkdtemp()
```
Done.

caisq commented on tensorboard/plugins/debugger/debugger_plugin_test.py:
```diff
@@ -0,0 +1,436 @@
+    value = event.summary.value.add(
+        tag=op_name,
+        node_name='%s:%d:DebugNumericSummary' % (op_name, output_slot))
+    value.tensor.tensor_shape.dim.add(size=len(list_of_values))
```
Good suggestion! Done.

caisq commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+    if (len(dim) != 1 or
+        dim[0].size < constants.MIN_DEBUG_NUMERIC_SUMMARY_TENSOR_LENGTH):
+      tf.logging.warning("Health-pill tensor either lacks a dimension or is "
+                         "size incorrectly. len(dim): %d, size of dim: %d",
```
Done.

caisq commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+      # Data from parameter servers (or any graphs without a master) do not
+      # contain core metadata. So the session run count is missing. Set its
+      # value to a microsecond epoch timestamp.
+      event.step = int(time.time() * 1e6)
```
It is okay. In the context of the debugger plugin, step counts don't have clear meanings. There can be multiple workers and parameter servers connecting to the plugin at the same time. Within each worker or ps, there can be concurrent tf.Session.run calls. We decided to use time stamp as a surrogate for the step count. This is not ideal, but will do the job under most cases.

wchargin replied:
Cool—as long as you say it's okay, I'm happy with it :-). Thanks for the explanation.

caisq commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+    # We create a special directory within logdir to store debugger-related
+    # events (if that directory does not already exist). This is necessary
+    # because for each directory within logdir, TensorBoard only reads through
+    # each events file once. There may be other non-debugger events files being
```
Ack.

caisq commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+    3) receives tensor value Event proto(s) through its on_value_event method.
+  """
+
+  def __init__(self,  # pylint: disable=super-init-not-called
```
The super class __init__ method currently raises a NotImplementedError. I'll fix it on the tensorflow side first. Added a TODO item for myself here.

caisq replied:
See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/debug/lib/grpc_debug_server.py#L68

wchargin replied:
Uh, wow. That's a bit special.

Yes, please do fix in TF. Thanks for explaining.

caisq commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+        constants.DEBUG_NUMERIC_SUMMARY_SUFFIX)
+
+    tensor = event.summary.value[0].tensor
+    dim = tensor.tensor_shape.dim
```
I agree. Done.

caisq commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+      receive_port: The port at which to receive health pills from the
+        TensorFlow debugger.
+      logdir: The directory in which to write events files that TensorBoard will
+        read. The ~ in this string might not be expanded yet.
```
I also prefer consistency. Removed this comment.

wchargin replied:
Great; thanks.

wchargin commented on .travis.yml:
```diff
@@ -42,6 +42,9 @@ before_install:
 
     pip install pylint
+    pip install futures==3.1.1
+    pip install grpcio==1.4.0
```
I'm skeptical as to whether this is the right way to declare dependencies. We already depend on several Python libraries from pypi—`bleach` and `markdown` among them—and we don't explicitly install any of them in the Travis config. (Installing `pylint` within Travis, on the other hand, makes sense, as it's only used as a separate test target.) I suspect that we instead want to add whichever packages end up being required to [third_party/python.bzl](https://github.com/tensorflow/tensorboard/blob/6fd136d8feaf11a81d1b287cc5029aad5e98ed85/third_party/python.bzl), and let Bazel do the dependency resolution automatically. (We don't do this for tensorflow and numpy, but they are exceptions, not the rule.) @jart please correct me if I'm wrong.

wchargin replied:
@caisq: Friendly ping on this one.

caisq replied:
I'm not sure whether it makes sense to build grpc/grpcio from source here. It is significantly larger than the dependencies in python.bzl. grpc source is about 4.3 MB, as compared with the sizes of html5lib, six and werkzeug sources, which are all below 400 kB. grpc source contains C++ and Python, while the aforementioned smaller ones are mostly python. So building grpc from source would likely significantly increase the build time. In this sense,  grpcio is more similar to the depencencies: numpy and tensorflow, in that it should be directly installed from binary for the build environment. When the debugger plugin is ready for a public release, grpcio should be listed as a dependency of the tensorboard pip package (or at least stated explicitly as a requirement in the error message if it is not installed.)

I'd like to hear @jart's opinion.

caisq commented on tensorboard/plugins/debugger/debugger_plugin_test.py:
```diff
@@ -0,0 +1,436 @@
+    self.debugger_data_server_grpc_port = 4242
+
+    # Fake threading behavior so that threads are synchronous.
+    class FakeThread(object):
```
This is actually not necessary anymore (it's a remnant from earlier versions of the unit tests). Removed it.

caisq replied:
Actually, I spoke too soon. Here we are patching the `start()` method of objects to be created in the future. I'm not sure if there is an easier way to do that.

wchargin replied:
My suggestion should work:
```py
Python 2.7.6 (default, Oct 26 2016, 20:30:19) 
[GCC 4.8.4] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> import threading
>>> def go():
...   print("I hail from %s" % threading.current_thread().name)
... 
>>> threading.Thread(name='my_new_thread', target=go).start()
I hail from my_new_thread
>>> stubs = tf.test.StubOutForTesting()
>>> stubs.Set(threading.Thread, 'start', threading.Thread.run)
>>> threading.Thread(name='my_new_thread', target=go).start()
I hail from MainThread
>>> stubs.CleanUp()
>>> threading.Thread(name='my_new_thread', target=go).start()
I hail from my_new_thread
```

caisq replied:
Cool. Thanks! I'm using this approach, but with tf.test.mock.patch for consistency with other places in this directory.

wchargin replied:
Fine with me!

wchargin commented on tensorboard/plugins/debugger/debugger_server_lib.py:
```diff
@@ -0,0 +1,321 @@
+    watch_key = event.summary.value[0].node_name
+    if not watch_key.endswith(constants.DEBUG_NUMERIC_SUMMARY_SUFFIX):
+      # Ignore events that lack a DebugNumericSummary.
+      # NOTE(chizeng): We may later handle other types of debug ops.
```
Apparently some internal security folks prefer `s/chizeng/@chihuahua` (#140).

caisq replied:
Done. Replaced existing instances of those with github usernames, including @chihuahua's and mine.

caisq commented on tensorboard/plugins/debugger/debugger_plugin_test.py:
```diff
@@ -0,0 +1,436 @@
+    shutil.rmtree(self.log_dir, ignore_errors=True)
+    mock.patch.stopall()
+
+  def _CreateEventWithDebugNumericSummary(
```
@chihuahua If I understand correctly, the debugger plugin is different from other plugins like histograms in that: 1) it writes summary files, but 2) the user doesn't write events explicitly. Instead, the plugin receives data from TF runtimes and store them in summary files. Do you think there is room for improvements in the way the summary data is created and tested here?

wchargin commented:
> [tdp] stands for [tensorboard debugger plugin]

*That* makes more sense! :-)

No need to use that label in this repository; the "TensorBoard" is assumed, and "debugger plugin" is clear from the rest of the commit title.

Thanks for clarifying!

caisq commented:
Thanks for the detailed and helpful reviewing so far, @wchargin!

wchargin commented:
You're welcome—thanks for your efforts on the integration!

caisq commented:
@wchargin, @chizeng, @jart 

This PR is updated after @chizeng pushed the changes to plugin data. All tests pass. I believe I've address ed all comments from @wchargin so far. 

PTAL

caisq commented:
Gentle ping @jart @wchargin @chihuahua

jart commented:
Out of curiosity, was this code reviewed internally? Do you have the CL numbers?

jart commented on .travis.yml:
```diff
@@ -42,6 +42,8 @@ before_install:
 
     pip install pylint
+    pip install grpcio==1.4.0
+    pip install portpicker==1.1.1
```
This dependency is so trivial that it shouldn't be difficult to add to our Bazel build.

caisq replied:
Done. Thanks for the suggestion.

jart commented on .travis.yml:
```diff
@@ -42,6 +42,8 @@ before_install:
     esac
 
     pip install pylint
```
Please `pip install futures`.

caisq replied:
Done.

jart commented on tensorboard/BUILD:
```diff
@@ -70,6 +71,14 @@ py_library(
 )
 
 py_library(
```
Please add an `expect_futures_installed`.

caisq replied:
Done.

jart commented on tensorboard/pip_package/setup.py:
```diff
@@ -25,6 +25,8 @@
 
 REQUIRED_PACKAGES = [
+    'futures >= 3.1.1',
+    'grpcio >= 1.4.0',
```
Please note that if we do this, it will become a transitive dependency of TensorFlow. Would that potentially cause problems?

caisq replied:
grpcio is fine as a transitive dependency of tensorflow. tensorflow is already using grpc (in its distributed_runtime, through C++). the size of grpcio is not excessive (at 4-5 MB), as compared with other dependency pip packages of tensorflow, e.g., numpy (15 MB) and protobuf (5 MB).

jart commented on tensorboard/plugins/debugger/debugger_plugin.py:
```diff
@@ -0,0 +1,504 @@
+    if self._debugger_data_server_grpc_port:
+      tf.logging.info('Creating DebuggerDataServer at port %d and logdir %s',
+                      self._debugger_data_server_grpc_port, self._logdir)
+      self._debugger_data_server = debugger_server_lib.DebuggerDataServer(
```
I strongly recommend against doing work in constructors. Consider exposing a `listen(PORT)` method instead, and calling it from the main function. Then remove the port field from this class and constructor.

caisq replied:
cc @chihuahua

I share your concern about starting some potentially heavy-weight thing in the constructor. However, I need some help understanding how what you said can be done. In main.py, `create_tb_app` takes a list of plugin constructors, instead of a list of actual plugin objects. So there is no way to call the proposed `listen` method in main.py. Maybe modify ``standard_tensorboard_wsgi` in `application.py` so it can call the `listen` method? But then there is the problem of how to pass the port argument in. Maybe through a flag? That feels a little messy. Let me know if there is a cleaner way to do it.

wchargin replied:
I'm not sure if this is what @jart means, but one approach: pass to `create_tb_app` a function that constructs a debugger plugin, invokes its `listen` method with a fixed port, and then returns the plugin object. Take advantage of the fact that the callable that you pass to `create_tb_app` need not be a "normal constructor."

Don't change `standard_tensorboard_wsgi` to call `listen`, no.

caisq replied:
@wchargin, good suggestion. Done.

jart commented on tensorboard/plugins/debugger/debugger_plugin.py:
```diff
@@ -0,0 +1,504 @@
+      A boolean. Whether this plugin is active.
+    """
+    return bool(
+        self._event_multiplexer and
```
Maybe return false if it isn't listening?

caisq replied:
Done.

caisq commented:
@jart re your question "Why not tf.test.mock?", the code you were referring to was out of date. It was already replaced with tf.test.mock.

caisq commented:
@jart I removed the thread- and server- startup from the constructor of `debugger_plugin.DebuggerPlugin` and added the `listen` method as you suggested. Thanks.

caisq commented:
Gentle ping @jart

jart commented on third_party/python.bzl:
```diff
@@ -71,3 +71,13 @@ def tensorboard_python_workspace():
+  native.new_http_archive(
+      name = "org_pythonhosted_portpicker",
+      urls = [
+          "https://pypi.python.org/packages/96/48/0e1f20fdc0b85cc8722284da3c5b80222ae4036ad73210a97d5362beaa6d/portpicker-1.1.1.tar.gz",
```
Please make `http://mirror.bazel.build/pypi.python.org/packages/96/48/0e1f20fdc0b85cc8722284da3c5b80222ae4036ad73210a97d5362beaa6d/portpicker-1.1.1.tar.gz` first in the list.

caisq commented:
@jart. Done adding the URL you suggested. Rebased and squashed. Thanks.

jart commented:
Will merge as soon as @wchargin's gives his nod.

caisq commented:
ping @wchargin

wchargin commented:
Sure—I'll take a look at this at most 60 minutes from now. (Trying to investigate an issue in our live release; give me a minute.)

caisq commented:
@wchargin. Comment addressed. I somehow missed that inline comment of yours. Sorry. But thanks again for the thorough review.

caisq commented:
@jart @wchargin : rebased and squashed again.

wchargin commented:
```
INFO: From Testing //tensorboard/plugins/debugger:debugger_plugin_test:
==================== Test output for //tensorboard/plugins/debugger:debugger_plugin_test:
Traceback (most recent call last):
  File "/home/travis/.bazel-output-base/bazel-sandbox/7099875188164452637/execroot/tensorboard/bazel-out/local-fastbuild/bin/tensorboard/plugins/debugger/debugger_plugin_test.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/debugger/debugger_plugin_test.py", line 29, in <module>
    from tensorboard.plugins.debugger import debugger_plugin_testlib
  File "/home/travis/.bazel-output-base/bazel-sandbox/7099875188164452637/execroot/tensorboard/bazel-out/local-fastbuild/bin/tensorboard/plugins/debugger/debugger_plugin_test.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/debugger/debugger_plugin_testlib.py", line 25, in <module>
    import portpicker
ImportError: No module named portpicker
```

caisq commented:
@wchargin The tests passed locally... Investigating.

jart commented on third_party/python.bzl:
```diff
@@ -71,3 +71,14 @@ def tensorboard_python_workspace():
   )
+
+  native.new_http_archive(
+      name = "org_pythonhosted_portpicker",
```
One last small request. Would it be possible to name this org_python_pypi_portpicker since portpicker has no affiliation with pythonhosted.org?

caisq replied:
Will do. I'm investigating why the new @org_pythonhosted_portpicker dependency doesn't get picked up by the new py_library and py_tests, leading to the ImportError.

caisq commented:
@jart @wchargin The previous error with `import portpicker` was due to a wrong strip_prefix for the portpicker tar ball. Fixed now. Rebased and squashed.
